dimred/src/gwas/train_val_split.py

```diff
@@ -1,224 +1,285 @@
+                part_phenotype = phenotype.merge(fold_indices, how='inner', on=['FID', 'IID'])
+
+                out_path = self.split.get_cov_pheno_path(node_index, fold_index, part)
+                part_phenotype.to_csv(out_path, sep='\t', index=False)
+
+        return out_path
+
+
+    def split_pca(self, node_index: int) -> str:
+        """
+        Extracts train or val subset of samples from file with principal components
+        Args:
+            node_index (int): Index of particular node
+
+        Returns:
+            str: Path to extracted subset of samples with PCs
+        """    
+        pca = pandas.read_table(self.split.get_source_pca_path(node_index))
+        pca.rename({'#FID': 'FID'}, axis='columns', inplace=True)
+
+        for fold_index in range(FOLD_COUNT):
+            for part in ['train', 'val', 'test']:
+                
+                fold_indices = pandas.read_table(self.split.get_ids_path(node_index, fold_index, part))
+                
+                fold_pca = pca.merge(fold_indices, how='inner', on=['FID', 'IID'])
+                fold_pca.to_csv(self.split.get_pca_path(node_index, fold_index, part), sep='\t', index=False)
+
+
+    def prepare_cov_and_phenotypes(self, node_index: int):
+        """Transforms phenotype+covariates and pca files into phenotype and pca+covariates files for each fold.
+
+        Args:
+            node_index (int): Index of node
+        """        
+        for fold_index in range(FOLD_COUNT):
+            for part in ['train', 'val', 'test']: 
+
+                pca_path = self.split.get_pca_path(node_index, fold_index, part)
+                cov_pheno_path = self.split.get_cov_pheno_path(node_index, fold_index, part)
+                pca_cov_path = self.split.get_pca_cov_path(node_index, fold_index, part)
+                phenotype_path = self.split.get_phenotype_path(node_index, fold_index, part)
+
+                self.prepare_cov_and_phenotype_for_fold(pca_path, cov_pheno_path, pca_cov_path, phenotype_path)
+
+
+    def prepare_cov_and_phenotype_for_fold(
+            self,
+            pca_path: str,
+            cov_pheno_path: str,
+            pca_cov_path: str,
+            phenotype_path: str
+        ):
+
+        """Transforms phenotype+covariates and pca files into phenotype and pca+covariates files.
+        This is required by plink 2.0 --glm command
+
+        Args:
+            pca_path (str): Path to PCA eigenvec file computed by plink 2.0
+            cov_pheno_path (str): Path to phenotype and covariates file prepared with ukb_loader
+            pca_cov_path (str): Path where all covariates including PCs will be stored
+            phenotype_path (str): Path with only phenotype data
+
+        """   
+        pca = pandas.read_table(pca_path)
+        cov_pheno = pandas.read_table(cov_pheno_path)
+        cov_columns = list(cov_pheno.columns)[2:-1]
```

Comment:
why not have a list of covariates and use column names rather than indices? Indices are not very robust + we might want to consider different sets of covariates

Reply:
In snakemake splitting version, we will use list of covariates as config parameters